AWS: Number of Opensearch requests are failing with wrong request signature

### Describe the bug

<!-- A clear and concise description of what the problem/bug is about and what is the current behavior. -->
When using AWS Opensearch some of he queries fail with wrong request signature error. 
1. When using AWS Opensearch with AWS S3 storage for snapshot repository, Operate (8.5) backup fails with exception:
```
Caused by: io.camunda.operate.exceptions.OperateRuntimeException: Exception occurred when validating whether backup with ID [<..>] already exists.
	at io.camunda.operate.webapp.opensearch.backup.OpensearchBackupRepository.validateNoDuplicateBackupId(OpensearchBackupRepository.java:160) ~[!/:8.5.0]
	at io.camunda.operate.webapp.backup.BackupService.takeBackup(BackupService.java:100) ~[!/:8.5.0]
	at io.camunda.operate.webapp.management.BackupController.takeBackup(BackupController.java:48) ~[!/:8.5.0]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:255) ~[spring-web-6.1.5.jar!/:6.1.5]
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:188) ~[spring-web-6.1.5.jar!/:6.1.5]
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:118) ~[spring-webmvc-6.1.5.jar!/:6.1.5]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:925) ~[spring-webmvc-6.1.5.jar!/:6.1.5]
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:830) ~[spring-webmvc-6.1.5.jar!/:6.1.5]
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87) ~[spring-webmvc-6.1.5.jar!/:6.1.5]
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089) ~[spring-webmvc-6.1.5.jar!/:6.1.5]
	... 120 more
Caused by: org.opensearch.client.opensearch._types.OpenSearchException: Request failed: [security_exception] The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

```

2. When starting Operate (8.4.6):
```
024-04-18 12:59:04.618 ERROR 8 --- [           main] i.c.o.s.SchemaStartup                    : Schema startup failed: Request failed: [security_exception] The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

The Canonical String for this request should have been
<..>
	at org.opensearch.client.transport.aws.AwsSdk2Transport.parseResponse(AwsSdk2Transport.java:479) ~[opensearch-java-2.5.0.jar!/:?]
	at org.opensearch.client.transport.aws.AwsSdk2Transport.executeSync(AwsSdk2Transport.java:393) ~[opensearch-java-2.5.0.jar!/:?]
	at org.opensearch.client.transport.aws.AwsSdk2Transport.performRequest(AwsSdk2Transport.java:191) ~[opensearch-java-2.5.0.jar!/:?]
	at io.camunda.operate.opensearch.ExtendedOpenSearchClient.arbitraryRequest(ExtendedOpenSearchClient.java:106) ~[operate-common-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.opensearch.ExtendedOpenSearchClient.arbitraryRequest(ExtendedOpenSearchClient.java:102) ~[operate-common-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.store.opensearch.client.sync.OpenSearchIndexOperations.lambda$getIndexNamesWithRetries$0(OpenSearchIndexOperations.java:44) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.store.opensearch.client.OpenSearchOperation.safe(OpenSearchOperation.java:39) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.store.opensearch.client.sync.OpenSearchIndexOperations.lambda$getIndexNamesWithRetries$2(OpenSearchIndexOperations.java:43) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.store.opensearch.client.sync.OpenSearchSyncOperation.withExtendedOpenSearchClient(OpenSearchSyncOperation.java:26) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.store.opensearch.client.sync.OpenSearchIndexOperations.getIndexNamesWithRetries(OpenSearchIndexOperations.java:42) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.schema.opensearch.OpensearchSchemaManager.getIndexNames(OpensearchSchemaManager.java:116) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.schema.IndexSchemaValidator.hasAnyOperateIndices(IndexSchemaValidator.java:109) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.schema.IndexSchemaValidator.validate(IndexSchemaValidator.java:87) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
	at io.camunda.operate.schema.SchemaStartup.initializeSchema(SchemaStartup.java:44) ~[operate-els-schema-8.4.6.jar!/:8.4.6]
<..>
2024-04-18 12:59:04.630  WARN 8 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'io.camunda.operate.schema.migration.SchemaMigration': Unsatisfied dependency expressed through field 'schemaStartup': Error creating bean with name 'schemaStartup': Invocation of init method failed
```


### To Reproduce

Steps to reproduce the behavior:
1. Configure the setup as described here: https://confluence.camunda.com/pages/viewpage.action?pageId=226133974
2. Start Operate of version 8.4.6 or 8.5.0 
3. or call Take backup endpoint on Operate 8.4.6

### Current behavior

<!-- Please attach screenshots, a screen recording, or a file (e.g. the BPMN/DMN/Form file) that has the problem you are describing to help us better debug the respective issue. -->
Backup fails with the error: `Exception occurred when validating whether backup with ID [<..>] already exists.`

### Expected behavior

<!-- A clear and concise description of what you expected to happen. -->
Backup is successfully created.

### Environment

<!-- Please provide details about the environment you were in when the problem occurred. -->
- Operate Versions: 8.4.6, 8.5.0
- Opensearch: 2.5
- AWS S3 storage

### Additional context

<!-- Please add any other context about the problem. Here you can also provide us some data that you used while the bug happen like **json** file or specific **BPMN**. -->
* Related support issues: https://jira.camunda.com/browse/SUPPORT-20972 and https://jira.camunda.com/browse/SUPPORT-21470
* Same works fine on Tasklist 8.4.6
* Probable problem is in [this part of code](https://github.com/camunda/zeebe/blob/afe4481e720510d108a847026da76de2426cb6b0/operate/common/src/main/java/io/camunda/operate/opensearch/ExtendedOpenSearchClient.java#L138), where we don't propagate request options. Comparing to Tasklist: [here](https://github.com/camunda/tasklist/blob/master/tasklist/webapp/src/main/java/io/camunda/tasklist/webapp/es/backup/os/BackupManagerOpenSearch.java#L364C44-L364C80).
* we have discrepancies in the method `OpenSearchIndexOperations#getIndexNamesWithRetries` in 8.4 and `main` branches, we need to clarify that

---------------------------------------------------------------------------------------------

### Test backup with AWS S3
1. Create a customized docker image for OpenSearch with AWS settings and plugin as this `Dockerfile`:
```
FROM opensearchproject/opensearch:2.5.0
    
ENV OPENSEARCH_INITIAL_ADMIN_PASSWORD StrongPassword1234!
ENV AWS_ACCESS_KEY_ID <INSERT-KEY-ID>
ENV AWS_SECRET_ACCESS_KEY <INSERT-ACCESS-KEY>
    
RUN /usr/share/opensearch/bin/opensearch-plugin install --batch repository-s3
RUN /usr/share/opensearch/bin/opensearch-keystore create
    
RUN echo $AWS_ACCESS_KEY_ID | /usr/share/opensearch/bin/opensearch-keystore add --stdin s3.client.default.access_key
RUN echo $AWS_SECRET_ACCESS_KEY | /usr/share/opensearch/bin/opensearch-keystore add --stdin s3.client.default.secret_key
    
# Optional
RUN echo $AWS_SESSION_TOKEN | /usr/share/opensearch/bin/opensearch-keystore add --stdin s3.client.default.session_token
```

2. Build the docker image and start as OpenSearch container
`docker build -t my-opensearch-s3 .`

3. Register s3 repository
`PUT http://localhost:9200/_snapshot/s3-backups-8.5`
```json
{
  "type": "s3",
  "settings": {
    "bucket": "<INSERT-BUCKET-NAME"
  }
}
```
4. Start Operate with repository name
`camunda.operate.backup.repositoryName=s3-backups-8.5`

5. Now you can test backup like described [here](https://docs.camunda.io/docs/8.4/self-managed/operational-guides/backup-restore/operate-tasklist-backup/) 
<!-- As the creator of the issue, you don't have to fill anything below this line, but the assignee will take care of this as part. -->

### Acceptance Criteria

<!-- the assignee will fill the Acceptance Criteria. -->

### Definition of Ready - Checklist

<!-- the assignee will check the DOR. -->
- [ ] The bug has been reproduced by the assignee in the environment compatible with the provided one; otherwise, the issue is closed with a comment
- [ ] The issue has a meaningful title, description, and testable acceptance criteria
- [ ] The issue has been labeled with an appropriate `Bug-area` label
- [ ] Necessary screenshots, screen recordings, or files are attached to the bug report

For UI changes required to solve the bug:

- [ ] Design input has been collected by the assignee

### Implementation

#### :mag: Root Cause Analysis

<!-- Explain the underlying cause for the issue. -->

#### :thought_balloon: Proposed Solution

<!-- Provide a high level overview of the proposed solution. (Technical details will be available in the associated PR) -->

### :point_right: Handover Dev to QA

<!--As a team, we have settled on a checklist to remind the DRI what information to provide to help the QA Engineer perform a frictionless and targeted QA test. The information requested by the checklist can be added before review/moving the ticket to the QA test column as a comment on the ticket.-->
- Changed components:
- Side effects on other components:
- Handy resources:
  BPMN/DMN models, plugins, scripts, REST API endpoints + example payload, etc :
  <!-- Add here -->
- Example projects:

<!-- Add here -->
- Commands/Steps needed to test; Versions to validate:

<!-- Add here -->
- Docker file / HELM chart : in case that it needed to be tested via docker share the version contain the fixed along with version of other services .

<!--elasticsearch: 16.2.2
identitiy:alpha3
zeebe:alpha3
Operate:alpha3
tasklist:alpha3-->
- Release version ( in which version this fixed/feature will be released):

<!-- Add here -->

### :green_book: Link to the test case

<!-- please add test case link for this bug if there is any if not after testing QA will  create a test case for it and add it here. -->

--- a/operate/common/src/main/java/io/camunda/operate/opensearch/ExtendedOpenSearchClient.java
+++ b/operate/common/src/main/java/io/camunda/operate/opensearch/ExtendedOpenSearchClient.java
@@ -47,12 +47,12 @@ public class ExtendedOpenSearchClient extends OpenSearchClient {
   private static final String DOCUMENT_ATTR =
       "org.opensearch.client:Deserializer:_global.search.TDocument";
 
-  public ExtendedOpenSearchClient(OpenSearchTransport transport) {
+  public ExtendedOpenSearchClient(final OpenSearchTransport transport) {
     super(transport);
   }
 
   private static <R> SimpleEndpoint<Map<String, Object>, R> arbitraryEndpoint(
-      String method, String path, JsonpDeserializer<R> responseParser) {
+      final String method, final String path, final JsonpDeserializer<R> responseParser) {
     return new SimpleEndpoint<>(
         request -> method, // Request method
         request -> path, // Request path
@@ -66,11 +66,11 @@ public class ExtendedOpenSearchClient extends OpenSearchClient {
     return ((JacksonJsonpMapper) transport.jsonpMapper()).objectMapper();
   }
 
-  private Map<String, Object> jsonToMap(String json) throws JsonProcessingException {
+  private Map<String, Object> jsonToMap(final String json) throws JsonProcessingException {
     return objectMapper().readValue(json, new TypeReference<>() {});
   }
 
-  private String json(SearchRequest request) {
+  private String json(final SearchRequest request) {
     final ByteArrayOutputStream baos = new ByteArrayOutputStream();
     final JsonGenerator generator = transport.jsonpMapper().jsonProvider().createGenerator(baos);
     request.serialize(generator, transport.jsonpMapper());
@@ -88,7 +88,7 @@ public class ExtendedOpenSearchClient extends OpenSearchClient {
   /*
 
   */
-  private String fixSearchAfter(String json) {
+  private String fixSearchAfter(final String json) {
     final Matcher m = SEARCH_AFTER_PATTERN.matcher(json);
     if (m.find()) {
       final var searchAfter = m.group(1); // Find "searchAfter" block in search request
@@ -103,7 +103,7 @@ public class ExtendedOpenSearchClient extends OpenSearchClient {
   }
 
   public <TDocument> SearchResponse<TDocument> fixedSearch(
-      SearchRequest request, Class<TDocument> tDocumentClass)
+      final SearchRequest request, final Class<TDocument> tDocumentClass)
       throws IOException, OpenSearchException {
     final var path = format("/%s/_search", join(",", request.index()));
     JsonEndpoint<Map<String, Object>, SearchResponse<Object>, ErrorResponse> endpoint =
@@ -117,24 +117,24 @@ public class ExtendedOpenSearchClient extends OpenSearchClient {
     return (SearchResponse<TDocument>) arbitraryRequest(requestJson, endpoint);
   }
 
-  public Map<String, Object> searchAsMap(SearchRequest request)
+  public Map<String, Object> searchAsMap(final SearchRequest request)
       throws IOException, OpenSearchException {
     final JsonEndpoint<SearchRequest, HashMap, ErrorResponse> endpoint =
         SearchRequest._ENDPOINT.withResponseDeserializer(getDeserializer(HashMap.class));
-
-    return transport.performRequest(request, endpoint, null);
+    return transport.performRequest(request, endpoint, transport.options());
   }
 
-  public Map<String, Object> arbitraryRequest(String method, String path, String json)
+  public Map<String, Object> arbitraryRequest(
+      final String method, final String path, final String json)
       throws IOException, OpenSearchException {
     final JsonEndpoint<Map<String, Object>, HashMap, ErrorResponse> endpoint =
-        arbitraryEndpoint(method, path, this.getDeserializer(HashMap.class));
+        arbitraryEndpoint(method, path, getDeserializer(HashMap.class));
     return arbitraryRequest(json, endpoint);
   }
 
   private <R> R arbitraryRequest(
-      String json, JsonEndpoint<Map<String, Object>, R, ErrorResponse> endpoint)
+      final String json, final JsonEndpoint<Map<String, Object>, R, ErrorResponse> endpoint)
       throws IOException, OpenSearchException {
-    return transport.performRequest(jsonToMap(json), endpoint, null);
+    return transport.performRequest(jsonToMap(json), endpoint, transport.options());
   }
 }

--- a/operate/common/src/test/java/io/camunda/operate/opensearch/ExtendedOpenSearchClientTest.java
+++ b/operate/common/src/test/java/io/camunda/operate/opensearch/ExtendedOpenSearchClientTest.java
@@ -1,0 +1,65 @@
+/*
+ * Copyright Camunda Services GmbH
+ *
+ * BY INSTALLING, DOWNLOADING, ACCESSING, USING, OR DISTRIBUTING THE SOFTWARE (“USE”), YOU INDICATE YOUR ACCEPTANCE TO AND ARE ENTERING INTO A CONTRACT WITH, THE LICENSOR ON THE TERMS SET OUT IN THIS AGREEMENT. IF YOU DO NOT AGREE TO THESE TERMS, YOU MUST NOT USE THE SOFTWARE. IF YOU ARE RECEIVING THE SOFTWARE ON BEHALF OF A LEGAL ENTITY, YOU REPRESENT AND WARRANT THAT YOU HAVE THE ACTUAL AUTHORITY TO AGREE TO THE TERMS AND CONDITIONS OF THIS AGREEMENT ON BEHALF OF SUCH ENTITY.
+ * “Licensee” means you, an individual, or the entity on whose behalf you receive the Software.
+ *
+ * Permission is hereby granted, free of charge, to the Licensee obtaining a copy of this Software and associated documentation files to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject in each case to the following conditions:
+ * Condition 1: If the Licensee distributes the Software or any derivative works of the Software, the Licensee must attach this Agreement.
+ * Condition 2: Without limiting other conditions in this Agreement, the grant of rights is solely for non-production use as defined below.
+ * "Non-production use" means any use of the Software that is not directly related to creating products, services, or systems that generate revenue or other direct or indirect economic benefits.  Examples of permitted non-production use include personal use, educational use, research, and development. Examples of prohibited production use include, without limitation, use for commercial, for-profit, or publicly accessible systems or use for commercial or revenue-generating purposes.
+ *
+ * If the Licensee is in breach of the Conditions, this Agreement, including the rights granted under it, will automatically terminate with immediate effect.
+ *
+ * SUBJECT AS SET OUT BELOW, THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * NOTHING IN THIS AGREEMENT EXCLUDES OR RESTRICTS A PARTY’S LIABILITY FOR (A) DEATH OR PERSONAL INJURY CAUSED BY THAT PARTY’S NEGLIGENCE, (B) FRAUD, OR (C) ANY OTHER LIABILITY TO THE EXTENT THAT IT CANNOT BE LAWFULLY EXCLUDED OR RESTRICTED.
+ */
+package io.camunda.operate.opensearch;
+
+import static org.assertj.core.api.Assertions.*;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyMap;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import java.io.IOException;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.opensearch.client.json.jackson.JacksonJsonpMapper;
+import org.opensearch.client.opensearch.core.SearchRequest;
+import org.opensearch.client.transport.Endpoint;
+import org.opensearch.client.transport.OpenSearchTransport;
+import org.opensearch.client.transport.TransportOptions;
+
+@ExtendWith(MockitoExtension.class)
+class ExtendedOpenSearchClientTest {
+
+  private @Mock OpenSearchTransport transport;
+  private ExtendedOpenSearchClient extendedClient;
+
+  @BeforeEach
+  void setUp() {
+    extendedClient = new ExtendedOpenSearchClient(transport);
+    assertThat(extendedClient).isNotNull();
+  }
+
+  @Test
+  void shouldUseTransportOptionsInArbitraryRequest() throws IOException {
+    when(transport.jsonpMapper()).thenReturn(new JacksonJsonpMapper());
+    when(transport.options()).thenReturn(mock(TransportOptions.class));
+    extendedClient.arbitraryRequest("GET", "/_snapshot/backups/camunda-part-1", "{}");
+    verify(transport).performRequest(anyMap(), any(Endpoint.class), any(TransportOptions.class));
+  }
+
+  @Test
+  void shouldUseTransportOptionsInSearchAsMap() throws IOException {
+    when(transport.options()).thenReturn(mock(TransportOptions.class));
+    extendedClient.searchAsMap(new SearchRequest.Builder().build());
+    verify(transport)
+        .performRequest(any(SearchRequest.class), any(Endpoint.class), any(TransportOptions.class));
+  }
+}

